safehttp/mux.go

```diff
@@ -184,38 +175,53 @@ func (s *ServeMuxConfig) Mux() *ServeMux {
 	if s.dispatcher == nil {
 		panic("Use NewServeMuxConfig instead of creating ServeMuxConfig using a composite literal.")
 	}
-	// pattern -> method -> handlerConfig
-	handlers := map[string]map[string]handlerConfig{}
+
+	methodNotAllowed := handlerConfig{
+		Dispatcher:   s.dispatcher,
+		Handler:      s.methodNotAllowedHandler.handler,
+		Interceptors: configureInterceptors(s.interceptors, s.methodNotAllowedHandler.cfgs),
+	}
+
+	handlers := map[string]*registeredHandler{}
 	for _, hr := range s.handlers {
 		if handlers[hr.pattern] == nil {
-			handlers[hr.pattern] = map[string]handlerConfig{}
-		}
-		if _, ok := handlers[hr.pattern][hr.method]; ok {
-			// TODO: this should be done in ServeMuxConfig.Handle, not here.
-			panic(fmt.Sprintf("double registration of (pattern = %q, method = %q)", hr.pattern, hr.method))
+			handlers[hr.pattern] = &registeredHandler{
+				pattern:          hr.pattern,
+				methodNotAllowed: methodNotAllowed,
+				methods:          make(map[string]handlerConfig),
+			}
 		}
-		handlers[hr.pattern][hr.method] =
+		handlers[hr.pattern].handleMethod(hr.method,
 			handlerConfig{
 				Dispatcher:   s.dispatcher,
 				Handler:      hr.handler,
 				Interceptors: configureInterceptors(s.interceptors, hr.cfgs),
-			}
-	}
-	methodNotAllowed := handlerConfig{
-		Dispatcher:   s.dispatcher,
-		Handler:      s.methodNotAllowedHandler.handler,
-		Interceptors: configureInterceptors(s.interceptors, s.methodNotAllowedHandler.cfgs),
+			})
 	}
 
 	mux := http.NewServeMux()
-	m := handlerMap(handlers, methodNotAllowed)
-	for pattern, handler := range m {
+	for pattern, handler := range handlers {
 		mux.Handle(pattern, handler)
 	}
-	return &ServeMux{mux: mux, handlerMap: m}
+	return &ServeMux{mux: mux, handlerMap: handlers}
+}
+
+// Clone creates a copy of the current config.
+// This can be used to create several instances of Mux that share the same set of
+// plugins and some common handlers.
+func (s *ServeMuxConfig) Clone() *ServeMuxConfig {
+	c := &ServeMuxConfig{
+		dispatcher:   s.dispatcher,
+		handlers:     make([]handlerRegistration, len(s.handlers)),
```

Comment:
nit: Instead of make + copy-later you can do this:

handlers: append([]handlerRegistration(nil), s.handlers...)

Reply:
Done.